<a href="https://colab.research.google.com/github/saadkamal219/Externals/blob/main/212_15_14729_DMML_Lab_Project_RHK.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

file_path = '/content/drive/MyDrive/Machine Learning Dataset/student_data.csv'

dataframe = pd.read_csv(file_path)

In [ ]:
dataframe.sample(5)

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
304     GP   M   19       U     GT3       T     3     3     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
277     GP   M   18       U     GT3       T     4     4   teacher  services   
188     GP   F   17       U     GT3       A     3     3    health     other   
53      GP   F   15       U     GT3       T     4     4  services  services   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
304  ...      4        4      4     1     1      3       20  15  14  13  
393  ...      4        4      1     3     4      5        0  11  12  10  
277  ...      3        2      4     1     4      3       22   9   9   9  
188  ...      3        3      3     1     3      3        6   8   7   9  
53   ...      3        3      4     2     3      5        0   8  10  11  

[5 rows x 33 columns]

In [ ]:
dataframe.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

In [ ]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

Encode the dataframe

In [ ]:
update_dataframe = dataframe.copy()

In [ ]:
from sklearn.preprocessing import OrdinalEncoder

ordinalencoder = OrdinalEncoder(categories=[['GP', 'MS'], ['M', 'F'], ['U', 'R'], ['GT3', 'LE3'], ['A', 'T'], ['at_home', 'health', 'services', 'teacher', 'other'], ['at_home', 'health', 'services', 'teacher', 'other']])
update_dataframe[['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob']] = ordinalencoder.fit_transform(update_dataframe[['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob']])

In [ ]:
update_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   school      395 non-null    float64
 1   sex         395 non-null    float64
 2   age         395 non-null    int64  
 3   address     395 non-null    float64
 4   famsize     395 non-null    float64
 5   Pstatus     395 non-null    float64
 6   Medu        395 non-null    int64  
 7   Fedu        395 non-null    int64  
 8   Mjob        395 non-null    float64
 9   Fjob        395 non-null    float64
 10  reason      395 non-null    object 
 11  guardian    395 non-null    object 
 12  traveltime  395 non-null    int64  
 13  studytime   395 non-null    int64  
 14  failures    395 non-null    int64  
 15  schoolsup   395 non-null    object 
 16  famsup      395 non-null    object 
 17  paid        395 non-null    object 
 18  activities  395 non-null    object 
 19  nursery     395 non-null    o

In [ ]:
dataframe['activities'].unique()

array(['no', 'yes'], dtype=object)

In [ ]:
ordinalencoder = OrdinalEncoder(categories=[['yes', 'no'], ['yes', 'no'], ['yes', 'no'], ['yes', 'no'], ['yes', 'no']])
update_dataframe[['romantic', 'internet', 'higher', 'nursery', 'activities']] = ordinalencoder.fit_transform(update_dataframe[['romantic', 'internet', 'higher', 'nursery', 'activities']])

In [ ]:
update_dataframe['reason'].unique()

array(['course', 'other', 'home', 'reputation'], dtype=object)

In [ ]:
update_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   school      395 non-null    float64
 1   sex         395 non-null    float64
 2   age         395 non-null    int64  
 3   address     395 non-null    float64
 4   famsize     395 non-null    float64
 5   Pstatus     395 non-null    float64
 6   Medu        395 non-null    int64  
 7   Fedu        395 non-null    int64  
 8   Mjob        395 non-null    float64
 9   Fjob        395 non-null    float64
 10  reason      395 non-null    object 
 11  guardian    395 non-null    object 
 12  traveltime  395 non-null    int64  
 13  studytime   395 non-null    int64  
 14  failures    395 non-null    int64  
 15  schoolsup   395 non-null    object 
 16  famsup      395 non-null    object 
 17  paid        395 non-null    object 
 18  activities  395 non-null    float64
 19  nursery     395 non-null    f

In [ ]:
ordinalencoder = OrdinalEncoder(categories=[['yes', 'no'], ['yes', 'no'], ['yes', 'no'], ['mother', 'father', 'other'], ['course', 'other', 'home', 'reputation']])
update_dataframe[['paid', 'famsup', 'schoolsup', 'guardian', 'reason']] = ordinalencoder.fit_transform(update_dataframe[['paid', 'famsup', 'schoolsup', 'guardian', 'reason']])

In [ ]:
update_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   school      395 non-null    float64
 1   sex         395 non-null    float64
 2   age         395 non-null    int64  
 3   address     395 non-null    float64
 4   famsize     395 non-null    float64
 5   Pstatus     395 non-null    float64
 6   Medu        395 non-null    int64  
 7   Fedu        395 non-null    int64  
 8   Mjob        395 non-null    float64
 9   Fjob        395 non-null    float64
 10  reason      395 non-null    float64
 11  guardian    395 non-null    float64
 12  traveltime  395 non-null    int64  
 13  studytime   395 non-null    int64  
 14  failures    395 non-null    int64  
 15  schoolsup   395 non-null    float64
 16  famsup      395 non-null    float64
 17  paid        395 non-null    float64
 18  activities  395 non-null    float64
 19  nursery     395 non-null    f

**Classifier**

In [ ]:
from sklearn.feature_selection import SelectKBest, f_classif

In [ ]:
x = update_dataframe.drop(['G3'], axis=1)

In [ ]:
y = update_dataframe['G3']

In [ ]:
selectkbest = SelectKBest(score_func=f_classif)
selectkbest.fit(x,y)

SelectKBest()

In [ ]:
statement = pd.DataFrame(selectkbest.scores_, index=x.columns, columns=['Score'])
selected_statement = statement.nlargest(32, 'Score')
selected_statement

Score
G2          145.931929
G1           90.677483
failures      6.115358
absences      3.810476
Walc          2.456415
schoolsup     1.989400
Dalc          1.952832
Medu          1.952109
romantic      1.834070
paid          1.807773
address       1.586767
goout         1.584186
higher        1.458866
age           1.397433
Fedu          1.317389
reason        1.205292
traveltime    1.193129
sex           1.188663
health        1.105824
Fjob          1.066927
studytime     1.062187
internet      1.055706
famsize       1.026348
Mjob          0.979960
Pstatus       0.965853
nursery       0.940990
activities    0.904321
freetime      0.848906
famsup        0.796305
guardian      0.687896
school        0.675116
famrel        0.598553

**Now we will select top 20 features based on the value of "Score" greater than and equal to 1.1**

In [ ]:
droplist = ['famrel', 'guardian', 'famsup', 'freetime', 'activities', 'nursery', 'Pstatus', 'Mjob', 'famsize', 'internet', 'Fjob']
selected_dataframe = update_dataframe.drop(droplist, axis=1)

In [ ]:
selected_dataframe.describe()

school         sex         age     address        Medu        Fedu  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     0.116456    0.526582   16.696203    0.222785    2.749367    2.521519   
std      0.321177    0.499926    1.276043    0.416643    1.094735    1.088201   
min      0.000000    0.000000   15.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000   16.000000    0.000000    2.000000    2.000000   
50%      0.000000    1.000000   17.000000    0.000000    3.000000    2.000000   
75%      0.000000    1.000000   18.000000    0.000000    4.000000    3.000000   
max      1.000000    1.000000   22.000000    1.000000    4.000000    4.000000   

           reason  traveltime   studytime    failures  ...      higher  \
count  395.000000  395.000000  395.000000  395.000000  ...  395.000000   
mean     1.440506    1.448101    2.035443    0.334177  ...    0.050633   
std      1.231309    0.697505    0.839240    0.743651  ...    0.219525   
min      0.000000    1.000000    1.000000    0.000000  ...    0.000000   
25%      0.000000    1.000000    1.000000    0.000000  ...    0.000000   
50%      2.000000    1.000000    2.000000    0.000000  ...    0.000000   
75%      3.000000    2.000000    2.000000    0.000000  ...    0.000000   
max      3.000000    4.000000    4.000000    3.000000  ...    1.000000   

         romantic       goout        Dalc        Walc      health    absences  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     0.665823    3.108861    1.481013    2.291139    3.554430    5.708861   
std      0.472300    1.113278    0.890741    1.287897    1.390303    8.003096   
min      0.000000    1.000000    1.000000    1.000000    1.000000    0.000000   
25%      0.000000    2.000000    1.000000    1.000000    3.000000    0.000000   
50%      1.000000    3.000000    1.000000    2.000000    4.000000    4.000000   
75%      1.000000    4.000000    2.000000    3.000000    5.000000    8.000000   
max      1.000000    5.000000    5.000000    5.000000    5.000000   75.000000   

               G1          G2          G3  
count  395.000000  395.000000  395.000000  
mean    10.908861   10.713924   10.415190  
std      3.319195    3.761505    4.581443  
min      3.000000    0.000000    0.000000  
25%      8.000000    9.000000    8.000000  
50%     11.000000   11.000000   11.000000  
75%     13.000000   13.000000   14.000000  
max     19.000000   19.000000   20.000000  

[8 rows x 22 columns]

**We selected 22 features for now because some features seemed important even the score in low**

**Algorithm Implementation**

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=0.6, random_state=3)

Attempt 1

In [ ]:
linearModel = LinearRegression()
linearModel.fit(x, y)

LinearRegression()

In [ ]:
y_pred = linearModel.predict(x_test)

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
import numpy as np

print("Mean Absolute Error:    ", mean_absolute_error(y_test, y_pred))
print("Mean squared Error:     ", mean_squared_error(y_test, y_pred))
print("Root Mean squared Error:", np.sqrt(mean_squared_error(y_test, y_pred)))

result = r2_score(y_test, y_pred)

print("Accuracy:               ", result*100)

Mean Absolute Error:     1.1324966926018483
Mean squared Error:      2.6665596310304607
Root Mean squared Error: 1.6329603886899586
Accuracy:                87.88811780210214


Attempt 2


In [ ]:
from sklearn import tree

dicission_tree = tree.DecisionTreeClassifier()

dicission_tree.fit(x_train, y_train)

DecisionTreeClassifier()

In [ ]:
y_pred_tree = dicission_tree.predict(x_test)

In [ ]:
print("Mean Absolute Error:    ", mean_absolute_error(y_test, y_pred_tree))
print("Mean squared Error:     ", mean_squared_error(y_test, y_pred_tree))
print("Root Mean squared Error:", np.sqrt(mean_squared_error(y_test, y_pred_tree)))

result = r2_score(y_test, y_pred_tree)

print("Accuracy:               ", result*100)

Mean Absolute Error:     1.1329113924050633
Mean squared Error:      4.2215189873417724
Root Mean squared Error: 2.0546335408879544
Accuracy:                80.82527760644386


In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=21)
knn.fit(x_train,y_train)

y_pred_knn = knn.predict(x_test)

print("Mean Absolute Error:    ", mean_absolute_error(y_test, y_pred_knn))
print("Mean squared Error:     ", mean_squared_error(y_test, y_pred_knn))
print("Root Mean squared Error:", np.sqrt(mean_squared_error(y_test, y_pred_knn)))

result = r2_score(y_test, y_pred_knn)

print("Accuracy:               ", result*100)

Mean Absolute Error:     1.4683544303797469
Mean squared Error:      5.6075949367088604
Root Mean squared Error: 2.3680360927800193
Accuracy:                74.52952917437669


In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=21)
rf.fit(x_train,y_train)

y_pred_rf = rf.predict(x_test)

print("Mean Absolute Error:    ", mean_absolute_error(y_test, y_pred_rf))
print("Mean squared Error:     ", mean_squared_error(y_test, y_pred_rf))
print("Root Mean squared Error:", np.sqrt(mean_squared_error(y_test, y_pred_rf)))

result = r2_score(y_test, y_pred_rf)

print("Accuracy:               ", result*100)

Mean Absolute Error:     1.4493670886075949
Mean squared Error:      7.19620253164557
Root Mean squared Error: 2.6825738632227014
Accuracy:                67.3138540307746


# After using multiple algorithms, we can see that Linear Regression comes with the highest accuracy 87.888% ~ 88%. So it is the final algorithm for my project
